rspace/src/main/scala/coop/rchain/rspace/pure/StoreInstances.scala

```diff
@@ -0,0 +1,411 @@
+                                            channelsHash,
+                                            newWaitingContinuation +: oldWaitingContinuations
+                                              .getOrElse(Seq.empty[WaitingContinuationBytes]))
+          })
+        })
+
+      override def removeJoin(txn: T, channel: C, channels: Seq[C]): ReaderT[F, LMDBContext, Unit] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            Option(ctx.dbJoins.get(txn, joinKey))
+              .map(toByteVectors)
+              .map(_.map(fromByteVectors[C]))
+              .map(exSeq => (exSeq, exSeq.indexOf(channels)))
+              .map {
+                case (exSeq, idx) =>
+                  if (idx >= 0) {
+                    val channelsHash = hashChannelsInner(channels)
+                    if (ctx.dbWaitingContinuations.get(txn, channelsHash) == null) {
+                      val resSeq = util.dropIndex(exSeq, idx)
+                      if (resSeq.nonEmpty) {
+                        ctx.dbJoins.put(txn, joinKey, toByteBuffer(resSeq.map(toByteVectorSeq(_))))
+                      } else {
+                        ctx.dbJoins.delete(txn, joinKey)
+                        collectGarbage(txn, joinKey, joinsCollected = true)
+                      }
+                    }
+                  } else {
+                    throw new IllegalArgumentException(
+                      s"removeJoin: $channels is not a member of $exSeq")
+                  }
+              }
+              .getOrElse(())
+        })
+
+      override def removeAllJoins(txn: T, channel: C): ReaderT[F, LMDBContext, Unit] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            ctx.dbJoins.delete(txn, joinKey)
+            collectGarbage(txn, joinKey)
+        })
+
+      override def putJoin(txn: T, channel: C, channels: Seq[C]): ReaderT[F, LMDBContext, Unit] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            val oldJoinsBv =
+              Option(ctx.dbJoins.get(txn, joinKey))
+                .map(toByteVectors)
+                .getOrElse(Seq.empty[Seq[ByteVector]])
+
+            val newJoin = toByteVectorSeq(channels)
+            if (!oldJoinsBv.contains(newJoin)) {
+              ctx.dbJoins.put(txn, joinKey, toByteBuffer(newJoin +: oldJoinsBv))
+            }
+        })
+
+      override def getJoin(txn: T, channel: C): ReaderT[F, LMDBContext, Seq[Seq[C]]] =
+        ReaderT(ctx =>
+          capture {
+            val joinKey = hashChannelsInner(Seq(channel))
+            Option(ctx.dbJoins.get(txn, joinKey))
+              .map(toByteVectors)
+              .map(_.map(fromByteVectors[C]))
+              .getOrElse(Seq.empty[Seq[C]])
+        })
+
+      override def removeAll(txn: T, channels: Seq[C]): ReaderT[F, LMDBContext, Unit] =
+        hashChannels(channels)
```

Comment:
If you...
```scala
import cats.syntax.applicative._
```
...then you can:
```scala
        for {
          channelsHash <- hashChannels(channels)
          mcs          <- readWaitingContinuationByteses(txn, channelsHash)
          _            <- writeWaitingContinuationByteses(txn, channelsHash, Seq.empty).whenA(mcs.isDefined)
          mds          <- readDatumByteses(txn, channelsHash)
          _            <- writeDatumByteses(txn, channelsHash, Seq.empty).whenA(mds.isDefined)
        } yield channels.map(c => removeJoin(txn, c, channels))
```

Reply:
For yet unknown reason whenA fails to compile with strange error messages. I will investigate it later. Currently just beautified existing implementation.